### PR TITLE
Validate incoming Record schemas

### PR DESCRIPTION
**Pull request by jeancochrane**
# Overview

In the course of reviewing #9, @ddohler noticed that we weren't actually validating that new Records matched their schema when creating them in the API; all of the validation logic was happening in the schema editor frontend.

This PR adds a validation step to Records to confirm that the new `data` field matches the associated RecordSchema.

## Notes

- Since the view tests and the Record validation code were both starting to get a bit bloated, I took this opportunity to do a small refactor: I split the Record validation tests out of `test_views.py` and into `test_record_validation.py`, and I reorganized `Record.clean()` to use descriptive methods (`clean_geom`, `clean_datetime`, `clean_data`). Apologies if this makes the diff harder to read!
- The commit history for this PR is not very clean, including one commit that introduces a failing test. As a result, I'll squash the PR before merging.

## Testing instructions

Travis will run the existing unit test suite for all supported versions of Python and Django. To run the tests locally:

```bash
# Build containers and run migrations.
./scripts/update

# Run tests.
./scripts/test app
```

Note that this will install virtualenvs for every version of Python and Django in the support matrix. If you'd like to remove virtualenvs after you're done, make sure to run `sudo rm -Rf .tox` in the project directory.

Closes https://github.com/azavea/grout-2018-fellowship/issues/44.

### CODE REVIEW

File: tests/test_record_validators.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        cls.record_schema = models.RecordSchema.objects.create(
+            record_type=cls.record_type,
+            version=1,
+            schema=cls.schema
+        )
+        # The base endpoint for listing records.
+        cls.record_endpt = reverse('record-list')
+
+    def test_temporal_record_with_no_datetime_info(self):
```
Comment (jeancochrane) [inline tests/test_record_validators.py:58]: This test was moved into this test module from `tests.test_views`.

File: tests/test_record_validators.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST, response.content)
+
+        min_msg = json.loads(response.content.decode('utf-8')).get('occurred_from')
+        max_msg = json.loads(response.content.decode('utf-8')).get('occurred_to')
+        expected_msg = exceptions.DATETIME_REQUIRED.format(uuid=self.record_type.uuid)
+
+        self.assertEqual(min_msg, expected_msg, response.content)
+        self.assertEqual(max_msg, expected_msg, response.content)
+
+    def test_occurred_from_greater_than_occurred_to(self):
```
Comment (jeancochrane) [inline tests/test_record_validators.py:80]: This test was also moved into this test module from `tests.test_views`.

File: grout/models.py [hunk 3]
```diff
@@ -112,40 +113,52 @@ class Record(GroutModel):
     class Meta(object):
         ordering = ('-created',)
 
-    def clean(self):
-        """
-        Provide custom field validation for this model.
+    def clean_geom(self):
```
Comment (ddohler) [inline grout/models.py:116]: Great refactor! 👍

**Review by ddohler — APPROVED** [review]
Looks good to me! My tests all passed locally but it might be a good idea to figure out why Travis is unhappy before merging.

**Comment by jeancochrane** [comment]
Looks like #10 introduced some test setup that didn't pass the new schema validation constraints. (Further verification that it works, at least!) I rebased off of develop and fixed the tests, so this is ready to roll.

Apologies for requesting your review before all tests passed, @ddohler! At the very least, this was a good lesson for me that I need to wait for both `pr` and `push` to pass before I can be confident all tests are passing.

**Pull request merged**